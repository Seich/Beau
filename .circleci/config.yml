version: 2
jobs:
    test:
        docker:
            - image: circleci/node:8.9.3
        steps:
            - checkout
            - run: npm install
            - run: npm run test:coverage
            - run: npx codeclimate-test-reporter < ./coverage/lcov.info
            - run: npm test
workflows:
    version: 2
    test:
        jobs:
            - test
