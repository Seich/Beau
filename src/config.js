const deepMerge = require('deepmerge');
const { requestRegex, UpperCaseKeys } = require('./shared');
const Plugins = require('./plugins');
const version = parseInt(require('../package.json').version, 10);

class Config {
    constructor(doc, env = {}) {
        const defaultConfigValues = {
            VERSION: version,
            ENDPOINT: '',
            PLUGINS: [],
            DEFAULTS: {},
            ENVIRONMENT: {},
            HOSTS: [],
            COOKIEJAR: false
        };

        this.configKeys = Object.keys(defaultConfigValues);

        let config = this.loadConfig(doc);
        Object.assign(this, defaultConfigValues, config);

        this.ENVIRONMENT = deepMerge(this.ENVIRONMENT, env);

        this.REQUESTS = [];

        this.loadRequests(doc, {
            DEFAULTS: this.DEFAULTS,
            ENDPOINT: this.ENDPOINT
        });

        this.loadHosts(this.HOSTS, config, defaultConfigValues);

        this.PLUGINS = new Plugins(this.PLUGINS);
    }

    loadHosts(hosts, rootConfig, defaultConfigValues) {
        hosts.forEach(host => {
            if (typeof host.host === 'undefined') {
                throw new Error(`Host doesn't indicate it's host name.`);
            }

            let config = deepMerge(defaultConfigValues, this.loadConfig(host));

            config.DEFAULTS = deepMerge(rootConfig.DEFAULTS, config.DEFAULTS);

            this.loadRequests(host, {
                DEFAULTS: config.DEFAULTS,
                ENDPOINT: config.ENDPOINT,
                NAMESPACE: host.host
            });
        });
    }

    loadRequests(host, settings) {
        let requests = Object.keys(host)
            .filter(key => requestRegex.test(key))
            .map(key => {
                let requestDefinitionIsString = typeof host[key] === 'string';
                let originalRequest = requestDefinitionIsString
                    ? { ALIAS: host[key] }
                    : host[key];

                let request = UpperCaseKeys(originalRequest);

                if (settings.NAMESPACE) {
                    request.ALIAS = `${settings.NAMESPACE}:${request.ALIAS}`;
                }

                request.REQUEST = key;
                request.COOKIEJAR = this.COOKIEJAR;
                request.ENDPOINT = settings.ENDPOINT;

                let defaults = UpperCaseKeys(settings.DEFAULTS);

                return deepMerge(defaults, request);
            });

        this.REQUESTS = this.REQUESTS.concat(requests);
    }

    loadConfig(host) {
        let config = {};

        Object.keys(host)
            .filter(k => this.configKeys.includes(k.toUpperCase()))
            .forEach(k => (config[k.toUpperCase()] = host[k]));

        return config;
    }
}

module.exports = Config;
