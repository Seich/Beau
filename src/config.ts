import { requestRegex, moduleVersion } from './shared'
import Plugins from './plugins'
import { safeLoad } from 'js-yaml'

const deepmerge = require('deepmerge')

export interface BeauConfig {
    cookiejar?: boolean
    defaults?: RequestObject
    endpoint?: string
    environment?: {
        [key: string]: UObjectString
    }
    host?: string
    plugins?: UObjectString[] | Plugins
    version?: number
    hosts?: BeauConfig[]
}

export interface RequestObject {
    endpoint?: string
    cookiejar?: boolean
    alias?: string
    form?: { [key: string]: any }
    formdata?: { [key: string]: any }
    headers?: { [key: string]: any }
    params?: { [key: string]: any }
    payload?: UObjectString
    request?: string
}

export type RequestConfig = RequestObject | RequestObject[] | string

export type UObjectString = { [key: string]: any } | string

export default class Config implements BeauConfig {
    version = moduleVersion()
    cookiejar = false
    endpoint = ''

    defaults: RequestObject = {}
    plugins: Plugins
    environment = {}
    host?: string = undefined
    hosts: BeauConfig[] = []

    requests: RequestObject[] = []

    constructor(config: BeauConfig) {
        this.version = config.version ?? moduleVersion()
        this.cookiejar = config.cookiejar ?? false
        this.endpoint = config.endpoint ?? ''
        this.defaults = config.defaults ?? {}
        this.environment = config.environment ?? {}
        this.host = config.host ?? undefined
        this.hosts = config.hosts ?? []
        this.plugins = new Plugins()

        if (Array.isArray(config.plugins)) {
            this.plugins = new Plugins(config.plugins)
        }

        this.loadRequests(config)
        this.loadHosts(this.hosts, this)
    }

    loadHosts(hosts: BeauConfig[], rootConfig: Config) {
        hosts.forEach((host) => {
            if (typeof host.host === 'undefined') {
                throw new Error(`Host doesn't indicate it's host name.`)
            }

            host = deepmerge(rootConfig, host)
            host.defaults = deepmerge(rootConfig.defaults, host.defaults)

            this.loadRequests(host)
        })
    }

    loadRequests(host: BeauConfig) {
        Object.entries(host)
            .filter(([key]) => requestRegex.test(key))
            .forEach(([key, rDefinition]: [string, RequestConfig]) => {
                if (Array.isArray(rDefinition)) {
                    rDefinition.forEach((req) =>
                        this.addRequest(key, req, host)
                    )
                } else {
                    this.addRequest(key, rDefinition, host)
                }
            })
    }

    addRequest(
        key: string,
        request: RequestObject | string,
        settings: BeauConfig
    ) {
        if (typeof request === 'string') {
            request = {
                alias: request
            }
        }

        if (settings.host) {
            request.alias = `${settings.host}:${request.alias}`
        }

        request.request = key
        request.cookiejar = this.cookiejar
        request.endpoint = settings.endpoint

        this.requests.push(deepmerge(settings.defaults, request))
    }
}

export function parseBeauConfig(str: string) {
    const doc = safeLoad(str) as BeauConfig
    return doc
}
