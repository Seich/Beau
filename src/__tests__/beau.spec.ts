import Beau from '../beau'
import { parseBeauConfig } from '../config'
const shared = require('../shared')

const requireg = require('requireg')
requireg.resolving = false

shared.moduleVersion = jest.fn().mockReturnValue(1)

describe(`Beau's config Loader`, () => {
    it('should load the config', () => {
        const doc = parseBeauConfig(`
            version: 1
            endpoint: 'http://example.com'

            defaults:
                headers:
                    authentication: hello
        `)

        const beau = new Beau(doc)
        expect(beau).toMatchSnapshot()
    })

    it(`should load the request list using the configuration`, () => {
        const doc = parseBeauConfig(`
            version: 1
            endpoint: 'http://example.com'

            GET /posts/1: get-post
            GET /user:
                alias: user
                headers:
                    hello: world
        `)

        const beau = new Beau(doc)
        expect(beau.requests).toMatchSnapshot()
    })

    it('should display a warning if the module version and the beau file version are different', () => {
        let stdout
        let spy = jest
            .spyOn(console, 'warn')
            .mockImplementation((val) => (stdout = val))

        shared.moduleVersion.mockReturnValue(2)

        new Beau({ version: 1 })
        expect(stdout).toEqual('This Beau file expects v1. You are using v2.')

        spy.mockReset()
        spy.mockRestore()
    })
})
