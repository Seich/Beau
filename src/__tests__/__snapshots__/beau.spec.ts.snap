// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Beau's config Loader. should load the config 1`] = `
Beau {
  "config": Config {
    "cookiejar": false,
    "defaults": Object {
      "headers": Object {
        "authentication": "hello",
      },
    },
    "endpoint": "http://example.com",
    "environment": Object {},
    "host": undefined,
    "hosts": Array [],
    "plugins": Plugins {
      "autoload": Array [
        "std",
      ],
      "context": Object {},
      "registry": Object {
        "dynamicValues": Array [],
        "postRequestModifiers": Array [],
        "preRequestModifiers": Array [],
      },
    },
    "requests": Array [],
    "version": 1,
  },
  "requests": RequestList {
    "cache": RequestCache {
      "$cache": Object {
        "env": Object {},
      },
    },
    "list": Array [],
  },
}
`;

exports[`Beau's config Loader. should load the request list using the configuration 1`] = `
RequestList {
  "cache": RequestCache {
    "$cache": Object {
      "env": Object {},
    },
  },
  "list": Array [
    Request {
      "alias": "get-post",
      "cookiejar": false,
      "dependencies": Set {},
      "endpoint": "http://example.com",
      "form": Object {},
      "formdata": Object {},
      "headers": Object {},
      "originalRequest": Object {
        "alias": "get-post",
        "cookiejar": false,
        "endpoint": "http://example.com",
        "request": "GET /posts/1",
      },
      "params": Object {},
      "path": "/posts/1",
      "payload": Object {},
      "plugins": Plugins {
        "autoload": Array [
          "std",
        ],
        "context": Object {},
        "registry": Object {
          "dynamicValues": Array [],
          "postRequestModifiers": Array [],
          "preRequestModifiers": Array [],
        },
      },
      "request": "GET /posts/1",
      "verb": "GET",
    },
    Request {
      "alias": "user",
      "cookiejar": false,
      "dependencies": Set {},
      "endpoint": "http://example.com",
      "form": Object {},
      "formdata": Object {},
      "headers": Object {
        "hello": "world",
      },
      "originalRequest": Object {
        "alias": "user",
        "cookiejar": false,
        "endpoint": "http://example.com",
        "headers": Object {
          "hello": "world",
        },
        "request": "GET /user",
      },
      "params": Object {},
      "path": "/user",
      "payload": Object {},
      "plugins": Plugins {
        "autoload": Array [
          "std",
        ],
        "context": Object {},
        "registry": Object {
          "dynamicValues": Array [],
          "postRequestModifiers": Array [],
          "preRequestModifiers": Array [],
        },
      },
      "request": "GET /user",
      "verb": "GET",
    },
  ],
}
`;
